doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -126,42 +133,72 @@ Type declarations
-                       : [ .. ]
-      ltac2_constructordef   : `uident` [ ( `ltac2_type` , ... , `ltac2_type` ) ]
-      ltac2_fielddef         : [ mutable ] `ident` : `ltac2_type`
-
-   Aliases are just a name for a given type expression and are transparently
+   .. insertprodn tac2typ_def tac2rec_field
+
+   .. prodn::
+      tac2typ_def ::= {? @tac2typ_prm } @qualid {? {| := | ::= } @tac2typ_knd }
+      tac2typ_prm ::= @typ_param
+      | ( {+, @typ_param } )
+      tac2typ_knd ::= @tac2type5
+      | [ {? {? %| } {+| @tac2alg_constructor } } ]
+      | [ .. ]
+      | %{ {? {+; @tac2rec_field } {? ; } } %}
+      tac2alg_constructor ::= @ident
+      | @ident ( {*, @tac2type5 } )
+      tac2rec_field ::= {? mutable } @ident : @tac2type5
+
+   :n:`:=`
+     Defines a type with a manifest
+
+   :n:`::=`
+     Defines an open variant type, a special kind of variant type whose constructors are not
+     statically defined, but can instead be extended dynamically. A typical example
+     is the standard `exn` type for exceptions. Pattern matching on open variants must always
+     include a catch-all clause. They can be extended with this form, in which case
+     :token:`tac2typ_knd` should be in the form :n:`[ {? {? %| } {+| @tac2alg_constructor } } ]`.
+
+     .. todo PR above: does it add an additional constructor for an existing o v type or
+        does it create both the type (and apparently the constructor?
+
+   Without :n:`{| := | ::= }`
+     Defines an abstract type for use representing data from OCaml.  Not for
+     end users.
+
+   Each production of :token:`tac2typ_knd` defines one of four possible kinds
+   of definitions, respectively: alias, variant, open variant and record types.
+
+   .. cleanup
+       .. productionlist:: coq
+          ltac2_typedef    : `ltac2_type`
+                           : [ `ltac2_constructordef` | ... | `ltac2_constructordef` ]
+                           : { `ltac2_fielddef` ; ... ; `ltac2_fielddef` }
+                           : [ .. ]
+          ltac2_constructordef   : `uident` [ ( `ltac2_type` , ... , `ltac2_type` ) ]
+          ltac2_fielddef         : [ mutable ] `ident` : `ltac2_type`
+
+   Aliases are names for a given type expression and are transparently
    unfoldable to it. They cannot be recursive. The non-terminal
-   :production:`uident` designates identifiers starting with an uppercase.
+   :token:`qualid` designates identifiers starting with an uppercase.
 
    Variants are sum types defined by constructors and eliminated by
    pattern-matching. They can be recursive, but the `rec` flag must be
    explicitly set. Pattern matching must be exhaustive.
 
+   .. todo PR next sentence: is that right?
+
+   Open variants can be extended with additional constructors using the `::=` form.
+
    Records are product types with named fields and eliminated by projection.
    Likewise they can be recursive if the `rec` flag is set.
```

Comment:
I guess mention of `with` for mutually recursive type definitions should be added.